ncrx/libncrx.c

```diff
@@ -713,17 +717,96 @@ static void terminate_msg_and_dict(struct ncrx_msg *msg)
 		msg->text[msg->text_len] = '\0';
 		msg->dict_len = len - msg->text_len - 1;
 		msg->dict++;
+	} else {
+		msg->dict_len = 0;
 	}
 }
 
 struct ncrx_msg *ncrx_next_msg(struct ncrx *ncrx)
 {
-	struct ncrx_msg *msg = msg_list_pop(&ncrx->retired_list);
+	struct ncrx_msg *start_msg, *next_msg;
+	uint64_t last_seen_cont_seq = 0;
 
-	if (msg)
-		terminate_msg_and_dict(msg);
+	start_msg = msg_list_peek(&ncrx->retired_list);
+	if (!start_msg) {
+		return NULL;
+	}
 
-	return msg;
+	last_seen_cont_seq = start_msg->seq;
+
+	/* If strlen(start_msg->text) is 0, this was already part of a CONT
+	 * sequence we merged so we don't need to consider it for further merging
+	 */
+	if (!ncrx->p.cont_merging_enabled || !strlen(start_msg->text)) {
+		msg_list_del(start_msg, &ncrx->retired_list);
+		terminate_msg_and_dict(start_msg);
+		return start_msg;
+	}
+
+	if (start_msg->rx_at_mono + ncrx->p.cont_msg_timeout > ncrx->now_mono) {
+		/* It's possible that any message might be appended to by a CONT
+		 * message, so wait a while to see if one will appear to merge.
+		 */
+		return NULL;
+	}
+
+	/* Ok, we've waited long enough for this packet. Grab all of the CONTs next
+	 * in the queue and merge them with this message.
+	 */
+	msg_list_del(start_msg, &ncrx->retired_list);
+	terminate_msg_and_dict(start_msg);
+
+	next_msg = msg_list_peek(&ncrx->retired_list);
+	if (next_msg && next_msg->cont && next_msg->seq == start_msg->seq + 1) {
+		/* We start with no space free due to using the memory allocated in
+		 * copy_msg, make it as large as possible so as to avoid repeatedly
+		 * having to realloc
+		 */
+		if (!(start_msg->text = realloc(start_msg->text, NCRX_CONT_REALLOC_SIZE))) {
+			return NULL;
```

Comment:
This leaks `start_msg` in the failure case. If you save/restore the original ->text pointer you can just return it here.